### PR TITLE
addon manager feature gates

### PR DESCRIPTION
**Pull request by haoqing0110**

### CODE REVIEW

File: pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go [hunk 1]
```diff
@@ -211,6 +217,18 @@ func (n *clusterManagerController) sync(ctx context.Context, controllerContext f
 		return removeClusterManagerFinalizer(ctx, n.clusterManagerClient, clusterManager)
 	}
 
+	// If AddOnManager is not enabled, remove related resources
+	if operatorapiv1.ComponentModeType(config.AddOnManagerComponentMode) != operatorapiv1.ComponentModeTypeEnable {
```
Comment (qiujian16) [inline pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go]: can we do it in reconciler? I do not think we should clean it in the level.

Comment (haoqing0110) [reply pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go]: yes, moved to reconcile().

File: pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go [hunk 2]
```diff
@@ -61,7 +62,7 @@ type clusterManagerController struct {
 
 type clusterManagerReconcile interface {
 	reconcile(ctx context.Context, cm *operatorapiv1.ClusterManager, config manifests.HubConfig) (*operatorapiv1.ClusterManager, reconcileState, error)
-	clean(ctx context.Context, cm *operatorapiv1.ClusterManager, config manifests.HubConfig) (*operatorapiv1.ClusterManager, reconcileState, error)
+	clean(ctx context.Context, cm *operatorapiv1.ClusterManager, config manifests.HubConfig, resource ...string) (*operatorapiv1.ClusterManager, reconcileState, error)
```
Comment (qiujian16) [inline pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go]: I do not think we should the change the signature

File: pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go [hunk 3]
```diff
@@ -149,6 +166,7 @@ func (c *runtimeReconcile) reconcile(ctx context.Context, cm *operatorapiv1.Clus
 func (c *runtimeReconcile) clean(ctx context.Context, cm *operatorapiv1.ClusterManager, config manifests.HubConfig) (*operatorapiv1.ClusterManager, reconcileState, error) {
 	// Remove All Static files
 	managementResources := []string{namespaceResource} // because namespace is removed, we don't need to remove deployments explicitly
+
```
Comment (qiujian16) [inline pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go]: can we call cleanResource here?

File: pkg/helpers/helpers.go [hunk 4]
```diff
@@ -728,6 +744,18 @@ func FindRelatedResourcesStatus(
 	return nil
 }
 
+func RemoveRelatedResourcesStatus(
+	relatedResourcesStatuses *[]operatorapiv1.RelatedResourceMeta,
+	relatedResource operatorapiv1.RelatedResourceMeta) {
+	var result []operatorapiv1.RelatedResourceMeta
+	for _, v := range *relatedResourcesStatuses {
+		if !reflect.DeepEqual(v, relatedResource) {
```
Comment (qiujian16) [inline pkg/helpers/helpers.go]: relatedResource is a plain struct that you can directly use == I think

File: pkg/helpers/helpers.go [hunk 5]
```diff
@@ -717,6 +719,20 @@ func SetRelatedResourcesStatuses(
 	}
 }
 
+func RemoveRelatedResourcesStatuses(
```
Comment (qiujian16) [inline pkg/helpers/helpers.go:711]: this needs tests. And are you sure we need this? My understanding is everytime we will update the whole relatedResource list.

Comment (haoqing0110) [reply pkg/helpers/helpers.go:711]: Hmm, I think the existing func `SetRelatedResourcesStatuses` and this new func `RemoveRelatedResourcesStatuses` are to add/remove obj to/from the relatedResource when applying/deleting resource. They are running before `UpdateClusterManagerRelatedResourcesFn` (which is to update the whole list you mentioned here). 

Will add more testing code.

Comment (haoqing0110) [reply pkg/helpers/helpers.go:711]: Add some UT and add relatedResources check in integration testing.

File: pkg/helpers/helpers.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/helpers/helpers.go]: fixed

File: pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go]: fixed

**Pull request merged**